File: app/src/main/java/net/ginteam/carmen/view/activity/MapActivity.java
package net.ginteam.carmen.view.activity;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.UiSettings;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.maps.android.clustering.ClusterManager;

import net.ginteam.carmen.R;
import net.ginteam.carmen.contract.MapContract;
import net.ginteam.carmen.model.company.CompanyModel;
import net.ginteam.carmen.presenter.MapPresenter;
import net.ginteam.carmen.utils.CompanyClusterRenderer;

public class MapActivity extends ToolbarActivity implements MapContract.View, OnMapReadyCallback {

    public static final String CATEGORY_ID_ARG = "category_id";

    private MapContract.Presenter mPresenter;

    private ClusterManager<CompanyModel> mClusterManager;
    private GoogleMap mGoogleMap;

    private MapView mMapView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_map);

        updateDependencies();

        mMapView.onCreate(savedInstanceState);

        mPresenter = new MapPresenter();
        mPresenter.attachView(this);

        mMapView.getMapAsync(this);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            onBackPressed();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void showGoogleMap(final LatLng userLocation) {
        mMapView.onResume();
        if (userLocation != null) {
            if (checkPermission()) {
                mGoogleMap.setMyLocationEnabled(true);
            }
            mMapView.post(new Runnable() {
                @Override
                public void run() {
                    animateToLocation(userLocation);
                }
            });
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mGoogleMap = googleMap;

        mPresenter.getLastUserLocation();

        mClusterManager = new ClusterManager<>(getContext(), mGoogleMap);
        mClusterManager.setRenderer(new CompanyClusterRenderer(getContext(), mGoogleMap, mClusterManager));
        mClusterManager.setOnClusterItemClickListener(new ClusterManager.OnClusterItemClickListener<CompanyModel>() {
            @Override
            public boolean onClusterItemClick(CompanyModel companyModel) {
                Log.d("Marker", "CLICK");
                return true;
            }
        });
        mGoogleMap.setOnCameraIdleListener(mClusterManager);
        mGoogleMap.setOnMarkerClickListener(mClusterManager);

        UiSettings googleMapUiSettings = mGoogleMap.getUiSettings();
        googleMapUiSettings.setCompassEnabled(false);
        googleMapUiSettings.setAllGesturesEnabled(true);
        googleMapUiSettings.setMyLocationButtonEnabled(false);
    }

    @Override
    public MapView getMapView() {
        return mMapView;
    }

    @Override
    public AppCompatActivity getActivity() {
        return this;
    }

    @Override
    public void showCityListView() {}

    @Override
    public void animateToLocation(LatLng location) {
        if (location != null) {
            mGoogleMap.animateCamera(
                    CameraUpdateFactory.newCameraPosition(
                            new CameraPosition.Builder()
                                    .target(location)
                                    .zoom(12)
                                    .build()));
        }
    }

    @Override
    public Context getContext() {
        return this;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mPresenter.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        mPresenter.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    private void updateDependencies() {
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_close_map);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        mMapView = (MapView) findViewById(R.id.google_map);
    }

    private boolean checkPermission() {
        return !(ActivityCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && ActivityCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED);
    }
}

File: app/src/main/kotlin/net/ginteam/carmen/kotlin/view/activity/map/MapActivity.kt
package net.ginteam.carmen.kotlin.view.activity.map

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.MenuItem
import com.google.android.gms.maps.model.LatLng
import net.ginteam.carmen.R
import net.ginteam.carmen.kotlin.contract.MapActivityContract
import net.ginteam.carmen.kotlin.interfaces.Filterable
import net.ginteam.carmen.kotlin.model.CategoryModel
import net.ginteam.carmen.kotlin.model.CompanyModel
import net.ginteam.carmen.kotlin.prepareFragment
import net.ginteam.carmen.kotlin.presenter.company.map.MapActivityPresenter
import net.ginteam.carmen.kotlin.view.activity.BaseLocationActivity
import net.ginteam.carmen.kotlin.view.activity.CompanyDetailsActivity
import net.ginteam.carmen.kotlin.view.activity.filter.FiltersActivity
import net.ginteam.carmen.kotlin.view.fragment.company.BaseCompaniesFragment
import net.ginteam.carmen.kotlin.view.fragment.company.map.MapCompaniesFragment
import net.ginteam.carmen.kotlin.view.fragment.company.map.MapCompanyFragment

class MapActivity : BaseLocationActivity <MapActivityContract.View, MapActivityContract.Presenter>(),
        MapActivityContract.View, BaseCompaniesFragment.OnCompanySelectedListener,
        MapCompaniesFragment.OnShowFiltersActivityListener {

    override var mPresenter: MapActivityContract.Presenter = MapActivityPresenter()

    private var mSelectedCategory: CategoryModel? = null
    private var mSelectedCompany: CompanyModel? = null

    private lateinit var mCurrentFragment: Fragment

    companion object {
        const val CATEGORY_ARGUMENT = "category"
        const val COMPANY_ARGUMENT = "company"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // if we receive category
        mSelectedCategory?.let {
            // then we need to fetch location
            // and show MapCompaniesFragment
            mPresenter.fetchUserLocation()
        }

        // if we receive company
        mSelectedCompany?.let {
            // then we need to show MapCompanyFragment
            // with only one company
            mCurrentFragment = MapCompanyFragment.newInstance(it)
            prepareFragment(R.id.main_fragment_container, mCurrentFragment)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        finish()
        return true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            FiltersActivity.FILTER_CONFIRM_REQUEST_CODE -> {
                if (resultCode == Activity.RESULT_OK) {
                    // set filters for map fragment
                    (mCurrentFragment as? Filterable)?.setFilterQuery(
                            data!!.getStringExtra(FiltersActivity.RESULT_FILTER_ARGUMENT)
                    )
                }
            }
            else -> {
                super.onActivityResult(requestCode, resultCode, data)
            }
        }
    }

    override fun onCompanySelected(company: CompanyModel) {
        val intent = Intent(getContext(), CompanyDetailsActivity::class.java)
        intent.putExtra(CompanyDetailsActivity.COMPANY_ARGUMENT, company)
        startActivity(intent)
    }

    override fun onShowFiltersActivity(category: CategoryModel) {
        val intent = Intent(getContext(), FiltersActivity::class.java)
        intent.putExtra(FiltersActivity.CATEGORY_ARGUMENT, category)
        startActivityForResult(intent, FiltersActivity.FILTER_CONFIRM_REQUEST_CODE)
    }

    override fun getLayoutResId(): Int = R.layout.activity_map_new

    override fun showCompaniesMapFragment(startLocation: LatLng) {
        mCurrentFragment = MapCompaniesFragment.newInstance(mSelectedCategory!!, startLocation)
        prepareFragment(R.id.main_fragment_container, mCurrentFragment)
    }

    override fun updateViewDependencies() {
        super.updateViewDependencies()

        supportActionBar!!.setHomeAsUpIndicator(R.drawable.ic_close_map)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
    }

    override fun updateDependencies() {
        super.updateDependencies()

        mSelectedCategory = intent.getSerializableExtra(CATEGORY_ARGUMENT) as CategoryModel?
        mSelectedCompany = intent.getSerializableExtra(COMPANY_ARGUMENT) as? CompanyModel?
    }
}
